Jt/interval sync

Jtang-1 opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->
Currently, when the [custom property](https://dimagi.atlassian.net/wiki/spaces/USH/pages/2146965443/USH+New+WebApps+Project+Setup+Guide#Application-Settings) `cc-sync-after-form` is configured, the restore will expire every 5 minutes. Upon a Formplayer request, the restore expiration is checked and if expired, a sync will be done. This resulted in requests that included a sync to have a slow response. This causes the user to perceive their action as being slow.

With this change, for apps that have the custom property `cc-sync-after-form` configured, an interval_sync-db request will be sent periodically (every 5 minutes).  The intent is so that the syncing will occur between user actions instead of during a user action. The syncing only occurs if the user has been active between the previous sync and current check to sync. This is prevent continuous syncing for inactive users.

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
[Jira Ticket](https://dimagi.atlassian.net/browse/USH-4063)
[Spec](https://docs.google.com/document/d/1owJma7M8TcxiitpqTDB6StkwMd--HPITku0rh-XBba4/edit?tab=t.0#heading=h.d7s2sycdfk92)
[Related Formplayer PR](https://github.com/dimagi/formplayer/pull/1544)

For tracking user activity, I originally set a listener for 'click' and 'keydown' and update the lastActivityTime. But instead, I decided to tie a user activity with Formplayer middleware. 

Another issue that I addressed in [99d62d1](https://github.com/dimagi/commcare-hq/pull/34161/commits/99d62d172cfce9341edee56a6df95299c3f68d13) is that the interval-sync could occur in between restore expiration. For example
- Time 0 min: Interval-sync starts
- Time 1 min: Synced occurred (triggered via Formplayer request). 
- Time 5 min: Interval-sync attempts to sync but restore is not expired yet
- Time 6 min: restore is expired
- Time 7 min: User performs action (triggering sync)

The two options are to 
- sync upon interval-sync regardless of restore expiration. The downside is that the user action could trigger the first sync, then the interval-sync would occur shortly after.
- Reset interval-sync such that its time 0 is the same as when the last sync occurs. The user action could still trigger the first sync.
- Remove check to sync upon user action and only rely on interval-sync. However, interval-sync is paused if the user is inactive. So the user action upon returning from inactivity (until the interval-sync occurs) would be based on a stale restore. This would not be acceptable.

The implementation on this PR went with option 2.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
No feature flag but is specific for custom property `cc-sync-after-form` configured

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
Locally tested. The impact will be for only apps that have 
### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
https://dimagi.atlassian.net/browse/QA-6151

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->
This PR is preferably deployed after Formplayer [PR](https://github.com/dimagi/formplayer/compare/jt/interval-sync?expand=1). Things won't break even if this is deployed before Formplayer but the request will error if Formplayer is not deployed.
- [X] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change

Jtang-1 committed:
feat: sets and updates the time of last user activity into session storage

Jtang-1 committed:
feat: sync should only occur if user has been active between the previous and current sync

Jtang-1 committed:
feat: sets periodic check to sync and sync if criterias are met.

It needs to check that the interval does not already exist, otherwise
multiple will be created

Jtang-1 committed:
feat: handle toggling off interval

Jtang-1 committed:
feat: gets sync frequency based on app custom property configuration and initiaties interval syncing

On Formplayer, when cc-sync-after-form custom property is configured, the restore
expires every 5 minutes and requires a sync. Because of that, 5 minute syncing interval
is also set here

Jtang-1 committed:
feat: starts the interval syncing if it isn't already started

this is initiated upon Formplayer query to match the previous behavior where
restore expiration is checked and sync is initiated during Formplayer requests

Jtang-1 committed:
feat: stop checking for sync after entering a form

Syncing during a form can result in mismatches between the form
start dataset and the current dataset, and introduces unnecessary churn.

Jtang-1 committed:
feat: plumb syncing to send request to syncing request to formplayer

Jtang-1 committed:
bug: the interval syncing should only occur when inside an app (not when in the app list)

The Formplayer endpoint requires an appId since restore expiration time period
can be configured on app level

Jtang-1 committed:
bug: needs to restart interval if a sync already occurs as part of another request

On Formplayer side, restoreExpiration is checked before performing sync.
This change makes the interval-sync time 0 match with time 0 of restoreExpiration.
Otherwise, interval-sync would likely be performed when the restore isn't expired leading
to a request associated with a user action to run the sync.

Jtang-1 committed:
Merge branch 'master' into jt/interval-sync

Jtang-1 committed:
bug: updates lastUserActivityTime upon formplayer query

Jtang-1 committed:
bug: interval should not be restarted after entering a form

restarting the interval will misalign timing between restoreExpiration
and sync-interval

Jtang-1 committed:
bug: handle cases where metaData is not included in response

Jtang-1 committed:
bug: accessing incorrect property

Jtang-1 committed:
refactor: no value in pulling out function

Jtang-1 committed:
bug: handle cases where properties do no exist

Jtang-1 committed:
style: use constants

nospame commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/app.js:
```diff
@@ -553,6 +553,40 @@ hqDefine("cloudcare/js/formplayer/app", function () {
+        complete = function (response) {
+            if (response.status === 'retry') {
+                FormplayerFrontend.trigger('retry', response, function () {
+                    options.data = JSON.stringify($.extend(true, {mustRestore: true}, data));
```
Since we're immediately stringifying, is the deep copy through `$.extend` necessary?

Jtang-1 replied:
what do you mean by a deep copy? Doesn't $.extend just merge two objects? I did remove `true` argument since recursive merging isn't needed [c4bd9ed](https://github.com/dimagi/commcare-hq/pull/34161/commits/c4bd9edb57e0d264fddc7160f9cda702ad6f18f6)

nospame replied:
I see us use `_.extend` more frequently -- there's just no option for deep copying. As far as I am aware it's identical to the updated version here, just underscore instead of jQuery.

nospame commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/app.js:
```diff
@@ -553,6 +553,40 @@ hqDefine("cloudcare/js/formplayer/app", function () {
+                }, gettext('Waiting for server progress'));
+            } else {
+                FormplayerFrontend.trigger('clearProgress');
+                CloudcareUtils.formplayerSyncComplete(response.responseJSON.status === 'error');
```
I'm not sure I understand this line. Doesn't this mean any completed sync will always show as an error?

nospame commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/app.js:
```diff
@@ -185,7 +185,7 @@ hqDefine("cloudcare/js/formplayer/app", function () {
 
     FormplayerFrontend.on('startForm', function (data) {
-        FormplayerUtils.stopIntervalSync();
+        FormplayerFrontend.permitIntervalSync = false;
```
Does this change basically mean that the next time there _would_ be an interval sync, it won't happen because this is set to false?

Jtang-1 replied:
Yes. We don't want to sync while inside a form (based on what i read in this [PR](https://github.com/dimagi/formplayer/pull/751)) 

But I also want to keep the 5 minute interval so I can't stop and restart the interval .

nospame commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/api.js:
```diff
@@ -93,7 +93,7 @@ hqDefine("cloudcare/js/formplayer/menus/api", function () {
                             }
 
-                            formplayerUtils.startIntervalSync(params.appId, parsedMenus.metaData.attemptRestore);
+                            formplayerUtils.startIntervalSync(params.appId, parsedMenus?.metaData?.attemptRestore);
```
I don't think I've seen this syntax before. I like it (although my IDE doesn't).

Jtang-1 replied:
my IDE doesn't either. I also just learned about it https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

nospame commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/utils/utils.js:
```diff
@@ -462,7 +462,7 @@ hqDefine("cloudcare/js/formplayer/utils/utils", function () {
         hqRequire(["cloudcare/js/formplayer/app"], function (FormplayerFrontend) {
             let currentApp = FormplayerFrontend.getChannel().request("appselect:getApp", appId),
-                customProperties = currentApp.attributes.profile.custom_properties;
+                customProperties = currentApp?.attributes?.profile.custom_properties || {};
```
Assuming the implication here is that if `attributes` exists, `profile` always does (so it's safe to access `custom_properties`)?

Jtang-1 replied:
good point, I'm pretty sure that's true but made it more defensive just in case [40d700c](https://github.com/dimagi/commcare-hq/pull/34161/commits/40d700c1802116dc4750a699ae115122f26d63da)

nospame reviewed (commented):
I think this generally makes sense with a few clarifying questions. I'll nitpick that I believe every `let` added here could be a `const` since they don't get redefined within their scope.

Jtang-1 committed:
bug: entering a form triggers `startForm` and `queryFormplayer`. permitIntervalSync needs to be end in a false value

Jtang-1 committed:
style: remove paranthesis, use const instead of let

Jtang-1 committed:
defensively handles if profile is null

Jtang-1 committed:
remove unneccessary recursive merge

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/utils/utils.js:
```diff
@@ -459,5 +459,17 @@ hqDefine("cloudcare/js/formplayer/utils/utils", function () {
+        function shouldSync() {
+            let currentTime = Date.now(),
+                lastUserActivityTime =  sessionStorage.getItem("lastUserActivityTime") || 0,
+                elapsedTimeSinceLastActivity = currentTime - (lastUserActivityTime);
```
[dd281c7](https://github.com/dimagi/commcare-hq/pull/34161/commits/dd281c74432e446f152d7ed557d36f217d7546cf)

AddisonDunn reviewed (commented):
Took an initial pass. I think I'm a little confused about when the sync timer should be reset.

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/utils/utils.js:
```diff
@@ -459,5 +460,48 @@ hqDefine("cloudcare/js/formplayer/utils/utils", function () {
+            const useAggressiveSyncTiming = (customProperties[constants.POST_FORM_SYNC] === "yes");
+            if (useAggressiveSyncTiming) {
+                // Sync frequency is synchronized with Formplayer's restore expiration
+                Utils.setEnableIntervalSync(true, FIVE_MINUTES_IN_MILLISECONDS);
```
you're right that `startIntervalSync` is called for every request that goes through `queryFormplayer`. But a new interval isn't created here. A check happens in `setEnableIntervalSync` [here](https://github.com/dimagi/commcare-hq/blob/40d700c1802116dc4750a699ae115122f26d63da/corehq/apps/cloudcare/static/cloudcare/js/formplayer/utils/utils.js#L492) that only creates a new interval if one doesn't already exist. If one already exists, then nothing changes (a new interval isn't created)

Jtang-1 replied:
Thanks for the suggestion, I think this seems clearer [50e2b5b](https://github.com/dimagi/commcare-hq/pull/34161/commits/50e2b5bdf6d69d33f263870bb820f01947fe7caa)

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/collections.js:
```diff
@@ -33,6 +33,7 @@ hqDefine("cloudcare/js/formplayer/menus/collections", function () {
             'type',
             'noItemsText',
             'dynamicSearch',
+            'metaData',
```
Yes, this was added previously by Shubham https://github.com/dimagi/formplayer/pull/1526

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/middleware.js:
```diff
@@ -32,6 +32,9 @@ hqDefine("cloudcare/js/formplayer/middleware", function () {
     };
+    var setLastUserActiviyTime = function () {
+        sessionStorage.setItem("lastUserActivityTime",  Date.now());
+    };
```
This middleware wraps the `API` [here](https://github.com/dimagi/commcare-hq/blob/3e1e893a647d9c60c2aa506d8fdc9d641649bc2d/corehq/apps/cloudcare/static/cloudcare/js/formplayer/router.js#L120) so it is used whenever an API method is called

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/utils/utils.js:
```diff
@@ -459,5 +460,48 @@ hqDefine("cloudcare/js/formplayer/utils/utils", function () {
+                customProperties = currentApp?.attributes?.profile?.custom_properties || {},
+                FIVE_MINUTES_IN_MILLISECONDS = 1000 * 60 * 5;
+            if (attemptedRestore) {
+                Utils.setEnableIntervalSync(false);
```
Ya this is pretty confusing. I tried to explain it a bit in the description's technical summary. 

As you said, for `navigate_menu` requests (and maybe some other's) the response would include `attemptRestore` true if a restore was performed.  If that happens, as you said, we want to restart the interval-sync timer. I do that by turning it off first (as you highlighted) then turning it back on in the following lines [here](https://github.com/dimagi/commcare-hq/blob/40d700c1802116dc4750a699ae115122f26d63da/corehq/apps/cloudcare/static/cloudcare/js/formplayer/utils/utils.js#L475). 

Would it be clearer for the interface of this function to name the parameter `restartInterval`? Even as is, i think it's confusing to have a `restart` parameter for a `start...` function but I can't think of a better way.

Jtang-1 replied:
How does this look https://github.com/dimagi/commcare-hq/pull/34161/commits/588ed4c45a8f7bacae1d1c191b71c55d7236b54a

Jtang-1 committed:
feat: don't display sync completion banner for interval-syncs

improved user experience per product feedback

Jtang-1 committed:
style: pass appId directly

Jtang-1 committed:
refactor: DRY

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/app.js:
```diff
@@ -552,6 +553,40 @@ hqDefine("cloudcare/js/formplayer/app", function () {
+        };
+        FormplayerUtils.setCrossDomainAjaxOptions(options);
+        $.ajax(options);
+    });
```
Ya i thought about it for a while but there with the `complete` being slightly different, I didn't see a useful way of extracting out functionality.

I did just make a get function for the [8721b46](https://github.com/dimagi/commcare-hq/pull/34161/commits/8721b466915607fd18cf8b599423712580f152ab). We could also make a helper function for 
```
     options = {
      url: formplayerUrl + "/interval_sync-db",
      data: JSON.stringify(data),
      complete: complete,
  };
  FormplayerUtils.setCrossDomainAjaxOptions(options);
  $.ajax(options);
```

but i don't think it simplifies much

Jtang-1 replied:
Were you thinking something like this [588ed4c](https://github.com/dimagi/commcare-hq/pull/34161/commits/588ed4c45a8f7bacae1d1c191b71c55d7236b54a)

Jtang-1 committed:
bug: syntax error

Jtang-1 committed:
Merge branch 'master' into jt/interval-sync

Jtang-1 committed:
refactor: DRY sync-db and interval_sync-db

Jtang-1 committed:
refactor: use start and stop helper functions for clarity

Jtang-1 committed:
refactor: clarify the function is exclusive to app property and clarify restarting interval

Jtang-1 committed:
bug: interval needs logic of when it should stop

otherwise, after it is set, it will continue running even after navigating
to an app where it the interval should not be set

Jtang-1 committed:
lint

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/utils/utils.js:
```diff
@@ -459,5 +460,60 @@ hqDefine("cloudcare/js/formplayer/utils/utils", function () {
+                        currentApp = FormplayerFrontend.getChannel().request("appselect:getApp", urlObject.appId),
+                        customProperties = currentApp?.attributes?.profile?.custom_properties || {},
+                        useAggressiveSyncTiming = (customProperties[constants.POST_FORM_SYNC] === "yes");
+                    if (!useAggressiveSyncTiming){
```
No that wasn't what I was concerned about. The bug I was seeing is:
Configuration:
- the user has access to App1 and App2
- only App2 is configured with "cc-sync-after-form" (which enables the interval sync)

Workflow:
1. Entering App2 (which initiates the interval sync)
2. Returning home then selecting App1
3. App1 will continue the interval sync even though this app does not have "cc-sync-after-form" enabled

But do let me know if i understood your comment and you have any additional thoughts about this.

AddisonDunn reviewed (approved):
Thank you for indulging all my questions and comments!

Jtang-1 committed:
style: use trigger since no return value is needed

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/utils/utils.js:
```diff
@@ -501,7 +501,7 @@ hqDefine("cloudcare/js/formplayer/utils/utils", function () {
                     }
                     if (shouldSync() && FormplayerFrontend.permitIntervalSync) {
-                        FormplayerFrontend.getChannel().request("interval_sync-db", urlObject.appId);
+                        FormplayerFrontend.trigger("interval_sync-db", urlObject.appId);
```
A key difference that i'm aware of is that `request` can return values while `trigger` (`events`) don't. 

Marionette documents principles of when to use [events](https://marionettejs.com/docs/master/backbone.radio.html#when-to-use-events) and when to use [requests](https://marionettejs.com/docs/master/backbone.radio.html#when-to-use-requests). I mainly decided to use trigger to follow what exists and because it doesn't return anything.

orangejenny committed:
Merge branch 'master' into jt/interval-sync

orangejenny committed:
Removed duplicate  directives introduced by merging master

orangejenny commented:
@Jtang-1 I updated this branch to reconcile it with https://github.com/dimagi/commcare-hq/pull/34223 - merging master and then tweaking a couple of files where you and I added `use strict` in slightly different places.

orangejenny committed:
Removed optional chaining syntax

orangejenny commented:
@Jtang-1 Added ff387e62b9e to remove the optional chaining

orangejenny committed:
Removed one more bit of optional chaining

orangejenny committed:
Updated customProperties to default to {} if profile exists but custom_properties does not

Jtang-1 committed:
bug:Updated customProperties to default to {} if profile exists but custom properties does not
…m_properties does not

Jtang-1 committed:
lint

orangejenny commented:
@AddisonDunn Optional chaining is a good practice in general, but in HQ, our js infrastructure doesn't support it - [this issue](https://github.com/requirejs/r.js/issues/971).

Jtang-1 commented:
Thank you for the review Addison!

Robert-Costello commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/utils/utils.js:
```diff
@@ -459,5 +460,67 @@ hqDefine("cloudcare/js/formplayer/utils/utils", function () {
+                lastUserActivityTime =  sessionStorage.getItem("lastUserActivityTime") || 0,
+                elapsedTimeSinceLastActivity = currentTime - lastUserActivityTime,
+                isInApp = Utils.currentUrlToObject().appId !== undefined;
+            if (elapsedTimeSinceLastActivity <= delayInMilliseconds && isInApp) {
```
should this be `elapsedTimeSinceLastActivity >= delayInMilliseconds`, like if more time has elapsed since last activity than the delay time, then we should sync?

Jtang-1 replied:
I can see how that's confusing - i thought the same thing as you looking back at it again. But I think this is correct. The logic here is that we should only sync if the user has been active within the last 5 minutes (`delayInMilliseconds`). If the user has not been active in the last 5 minutes, then we don't want to sync. This to prevent continuously syncing if a user has web apps open but it's in the background and not actively used.

Robert-Costello replied:
ahh, ok. Thanks for clarifying!

Robert-Costello commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/utils/utils.js:
```diff
@@ -459,5 +460,67 @@ hqDefine("cloudcare/js/formplayer/utils/utils", function () {
 
+    Utils.setSyncInterval = function (appId, restartInterval) {
+        hqRequire(["cloudcare/js/formplayer/app"], function (FormplayerFrontend) {
+            const currentApp = FormplayerFrontend.getChannel().request("appselect:getApp", appId);
```
Where does this request go? I don't see a reply function for it, but maybe I'm missing something. Locally I swapped in `currentApp = AppsAPI.getAppEntity(appId)` and that seemed to work, but otherwise `currentApp` kept coming back as `undefined`

Jtang-1 replied:
Looks like `appselect:getApp` was removed in this commit `3b5f59be26a17d3052163ddf9c42c546fe1b4db1` and you're right, that this line should have been updated to `currentApp = AppsAPI.getAppEntity(appId)`

Robert-Costello replied:
That explains it. I can make a PR for that. I'm investigating some QA bugs for case search, so that's why I'm looking through old PR's.
